src/lib/modules/specialconfigs/index.js

```diff
@@ -133,54 +143,131 @@ class SpecialConfigs {
+            self.logger.error(err.message || err);
+            params.shouldDeploy = false;
+          } else if (!result) {
+            self.logger.info(params.contract.className + ' deployIf directive returned false; contract will not deploy');
+            params.shouldDeploy = false;
+          }
+
+          cb(params);
+        });
+      }
+    });
+  }
+
+  getOnDeployLifecycleHookDependencies(contractConfig) {
+    let dependencyNames = contractConfig.deps || [];
+    dependencyNames.push(contractConfig.className);
+
+    return new Promise((resolve, reject) => {
+      async.map(dependencyNames, (contractName, next) => {
+        this.events.request('contracts:contract', contractName, (contractRecipe) => {
+          if (!contractRecipe) {
+            next(new Error('ReferredContractDoesNotExist'));
+          }
+          this.events.request('blockchain:contract:create', {
+            abi: contractRecipe.abiDefinition,
+            address: contractRecipe.deployedAddress
+          }, contractInstance => {
+            next(null, { className: contractRecipe.className, instance: contractInstance });
+          });
+        });
+      }, (err, contractInstances) => {
         if (err) {
-          self.logger.error(params.contract.className + ' deployIf directive has an error; contract will not deploy');
-          self.logger.error(err.message || err);
-          params.shouldDeploy = false;
-        } else if (!result) {
-          self.logger.info(params.contract.className + ' deployIf directive returned false; contract will not deploy');
-          params.shouldDeploy = false;
+          reject(err);
         }
+        this.events.request('blockchain:get', web3 => resolve(this.assembleLifecycleHookDependencies(contractInstances, web3)));
+      });
+    });
+  }
 
-        cb(params);
+  getAfterDeployLifecycleHookDependencies() {
+    return new Promise((resolve, reject) => {
+      this.events.request('contracts:list', (err, contracts) => {
+        async.map(contracts, (contract, next) => {
+          this.events.request('blockchain:contract:create', {
+            abi: contract.abiDefinition,
+            address: contract.deployedAddress
+          }, contractInstance => {
+            next(null, { className: contract.className, instance: contractInstance });
+          });
+        }, (err, contractInstances) => {
+          if (err) {
+            reject(err);
+          }
+          this.events.request('blockchain:get', web3 => resolve(this.assembleLifecycleHookDependencies(contractInstances, web3)));
+        });
       });
     });
   }
 
+  assembleLifecycleHookDependencies(contractInstances, web3) {
+    let dependencies = { contracts: {}, web3 };
+
+    contractInstances.forEach(contractInstance => {
+      dependencies.contracts[contractInstance.className] = contractInstance.instance;
+    });
+    return dependencies;
+  }
```

Comment:
@alaibe this is the handler that's shared in both `get*Dependencies()` functions.

Notice that that's the only part that could potentially be shared.

Reply:
Fair enough, can I suggest using `reduce`? That will be more declarative